### PR TITLE
Task 20

### PR DESCRIPTION
**Pull request by artsiomMandrukevich**
# Description
Task20. Created 4 tests for validating `GET /zipcode` and `POST /zipcode/expand`.

## How to test / repro
Run `ZipCodesTest` class.

### Steps for reproducing:
1. Up the web-service in the docker container.
`docker run -d -p 7072:8082 coherentsolutions/rest-training:1.0`
2. Fire `POST /zip-codes/expand` endpoint with the following body: 
`["11103", "11103", "12345"]`
### Actual Result:
Duplicated zipCodes (available/used) were created into DB.
image

### Expected Result:
The web-service logic doesn't allow creating a duplicate zipCodes in the DB.

### CODE REVIEW

**Review by anton-pas — APPROVED** [review]
Nice work!

File: src/main/java/com/coherensolutions/rest/training/http/Client.java [hunk 1]
```diff
@@ -13,4 +17,23 @@ public CloseableHttpResponse sendPost(CloseableHttpClient httpClient, String URL
         return httpClient.execute(httpPost);
     }
 
+    @SneakyThrows
+    public CloseableHttpResponse sendGet(String URL, String bearerToken) {
```
Comment (KirKarpov) [inline src/main/java/com/coherensolutions/rest/training/http/Client.java]: Variable name should be in lowercase

File: src/main/java/com/coherensolutions/rest/training/http/Client.java [hunk 2]
```diff
@@ -13,4 +17,23 @@ public CloseableHttpResponse sendPost(CloseableHttpClient httpClient, String URL
...
+        CloseableHttpClient httpClient = HttpClients.createDefault();
+        HttpGet httpGet = new HttpGet(URL);
+        httpGet.setHeader(HttpHeaders.AUTHORIZATION, "Bearer" + bearerToken);
+        return httpClient.execute(httpGet);
+    }
+
+    @SneakyThrows
+    public CloseableHttpResponse sendPost(String URL, String bearerToken, StringEntity jsonBody) {
```
Comment (KirKarpov) [inline src/main/java/com/coherensolutions/rest/training/http/Client.java]: Variable name should be in lowercase

File: src/main/java/utils/Handler.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        return new StringEntity(listBody.toString());
+    }
+
+    @SneakyThrows
+    public List<String> getListFromResponse(CloseableHttpResponse response) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        String jsonBody = EntityUtils.toString(response.getEntity());
+        return objectMapper.readValue(jsonBody, new TypeReference<>() {
+        });
+    }
```
Comment (KirKarpov) [inline src/main/java/utils/Handler.java]: Those methods could be static. In such case you won't need to create instance of class to call method(s)

File: src/main/java/com/coherensolutions/rest/training/http/ClientZipCodes.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @SneakyThrows
+    public ZipCodesResponse sendPostZipCodes(List<String> listOfZipCodes, int statusCode) {
+        CloseableHttpResponse response = client.sendPost(urlPostZipCodes, token.getWriteToken(), handler.convertListIntoJsonBody(listOfZipCodes));
+        return getListZipFromBody(response, statusCode);
+    }
+
+    @SneakyThrows
+    public ZipCodesResponse getListZipFromBody(CloseableHttpResponse response, int statusCode) {
```
Comment (KirKarpov) [inline src/main/java/com/coherensolutions/rest/training/http/ClientZipCodes.java]: It is better to make method private

File: src/test/java/ZipCodesTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.List;
+import static org.assertj.core.api.Assertions.*;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class ZipCodesTest {
+
+    @Order(1)
+    @DisplayName("Scenario #1 from the task 20")
```
Comment (KirKarpov) [inline src/test/java/ZipCodesTest.java]: It is better to write information regarding what test is doing . For example, get all available zip codes

File: src/test/java/ZipCodesTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    @Test()
+    void zipCodesGetTest() {
+        List<String> testZipCodes = new ArrayList<>(Arrays.asList("12345" ,"23456", "ABCDE"));
```
Comment (KirKarpov) [inline src/test/java/ZipCodesTest.java]: Usually such variable are named by starting from expected word. For example, expectedZipCodes

**Review by KirKarpov — APPROVED** [review]

File: src/main/java/com/coherensolutions/rest/training/http/Client.java (hunk 1 above)
Comment (artsiomMandrukevich) [reply src/main/java/com/coherensolutions/rest/training/http/Client.java]: done

File: src/main/java/com/coherensolutions/rest/training/http/ClientZipCodes.java (hunk 4 above)
Comment (artsiomMandrukevich) [reply src/main/java/com/coherensolutions/rest/training/http/ClientZipCodes.java]: done

File: src/main/java/utils/Handler.java (hunk 3 above)
Comment (artsiomMandrukevich) [reply src/main/java/utils/Handler.java]: done

File: src/test/java/ZipCodesTest.java (hunk 6 above)
Comment (artsiomMandrukevich) [reply src/test/java/ZipCodesTest.java]: done

File: src/test/java/ZipCodesTest.java (hunk 5 above)
Comment (artsiomMandrukevich) [reply src/test/java/ZipCodesTest.java]: done

**Pull request merged**